instrumentation/kafka-clients/src/main/java/brave/kafka/clients/KafkaTracing.java

```diff
@@ -43,16 +43,28 @@ public static KafkaTracing create(Tracing tracing) {
   /**
    * Retrieve the span extracted from the record headers. Creates a root span if the context is not
    * available.
+   *
+   * @deprecated this results in appending to a span already complete. Please use {@link
+   * #nextSpan(ConsumerRecord)}
    */
+  @Deprecated
   public Span joinSpan(ConsumerRecord record) {
```

Comment:
@ImFlog found a bug.. we shouldn't encourage this (better remove or repurpose) as this will lead to writing to a finished span. I also noticed we could probably make a single span for the batch consumption (which could lead to significantly less load) cc @jonathan-lo

Reply:
Aren't we heading to the first problem i.e. how do we know when we start the record treatment ?

Reply:
so in messaging span, we have something defined like this:
[producer span], [consumer span], [processor span]

I think we (me) accidentally missed reviewing this impl, that "joinSpan" is too late. It would do something like this..
[producer span], [consumer span{end}       {late data}]

So the first problem is to not expose something that will always mutate a completed span when a trace is in progress, nextSpan accomplishes this by creating a separate span for this activity.

The other thing I noticed was that the "consumer span" could be no problem if used for the bulk task, as it would be the single parent of each processor span.
Ex  after we fix the joinSpan -> nextSpan, for a 2 message batch we have
```
[producer span] -> [consumer span] -> [processor span]
                   [consumer span] -> [processor span]
```
Seems we could make this more efficient by removing the message key from the consumer span (processor could add that), and then re-use one consumer span for the batch, like this.

```
[producer span] -> [consumer span] -> [processor span]
                                  \-> [processor span]
```

If there are problems with this, I currently don't see them. Can you reply if there are (and reply generally as you are key in this decision!)

Reply:
moved discussion here

https://github.com/openzipkin/brave/pull/518